corehq/apps/app_manager/views/utils.py

```diff
@@ -605,26 +612,74 @@ def get_cleaned_session_endpoint_id(module_or_form, raw_endpoint_id, app):
 
 
 def set_session_endpoint(module_or_form, raw_endpoint_id, app):
-    cleaned_id = get_cleaned_session_endpoint_id(module_or_form, raw_endpoint_id, app)
+    cleaned_id = get_cleaned_and_deduplicated_session_endpoint_id(module_or_form, raw_endpoint_id, app)
     module_or_form.session_endpoint_id = cleaned_id
 
 
+def set_shadow_module_and_form_session_endpoint(
+    shadow_module,
+    raw_endpoint_id,
+    form_session_endpoints,
+    app
+):
+    cleaned_module_session_endpoint_id = \
+        get_cleaned_session_endpoint_id(raw_endpoint_id)
+    cleaned_form_session_endpoint_ids = \
+        [get_cleaned_session_endpoint_id(m['session_endpoint_id']) for m in form_session_endpoints]
+
+    duplicate_ids = _duplicate_endpoint_ids(
+        cleaned_module_session_endpoint_id, cleaned_form_session_endpoint_ids, shadow_module.unique_id, app)
+
+    if len(duplicate_ids) > 0:
+        duplicates = ", ".join([f"'{d}'" for d in duplicate_ids])
+        raise AppMisconfigurationError(_(
+            f"Session endpoint IDs must be unique. {duplicates} are used multiple times"
+        ).format(duplicates=duplicates))
+
+    shadow_module.session_endpoint_id = raw_endpoint_id
+    shadow_module.form_session_endpoints = [
+        ShadowFormEndpoint(form_id=m['form_id'], session_endpoint_id=m['session_endpoint_id'])
+        for m in form_session_endpoints if m['session_endpoint_id']
+    ]
```

Comment:
I noticed that earlier in this function we clean `raw_endpoint_id` and the `session_endpoint_id`s in `form_session_endpoints`, but then we save the raw versions.  @MartinRiese do you recall why that is?  Shouldn't this save the cleaned values?

Reply:
@esoergel They should really be the same once you get past `get_cleaned_session_endpoint_id` since it raises an error if the cleaned one does not match the raw

Reply:
Ah right, that makes sense, thank you.